### PR TITLE
Merging NuGet.Protocol

### PR DESCRIPTION
**Pull request by emgarten**
Creating NuGet.Protocol from the following projects:
NuGet.Protocol.Core.v3
NuGet.Protocol.Core.Types
NuGet.Protocol.VisualStudio

Changes here are all reactions to the name changes, updating project references, and include lists.

I'd be happy to hear feedback on if any of these should not be combined, or if this should be combined in a different way.

//cc @rrelyea @joelverhagen

### CODE REVIEW

**Review by joelverhagen — APPROVED** [review]
👍 👌

File: src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+#if IS_DESKTOP
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: Can this file be conditionally compiled instead?

Comment (mishra14) [inline src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: Why is this needed here? Did any behavior change?

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: I did that first and it worked, but then moved to this. I'm open to either way.

My current thinking is that if defs are more obvious when working in Visual Studio on the code. I've seen VS have issues in the past with evaluating conditions such as the IsDesktop property we have on items to make this work.

Also, in VS you get the switcher drop down at the top to go back and forth. With having this in the csproj you wouldn't get that.

If defs are ugly however, so I can understand wanting this in the csproj.

Thoughts?

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: For the test projects which will be merged next there will be a lot more scenarios like this. So we should figure this out.

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: It was merged in from NuGet.Protocol.VisualStudio which was a net45 project. NuGet.Protocol is a net45/netstandard1.3 project. 

The desktop framework requirement comes from using MEF here, this class is only used in the VSIX.

Comment (alpaix) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: The entire `NuGet.Protocol.VisualStudio` was built on `net45` only. At the same time I couldn't find direct dependencies of it in any other projects (besides some test project). The only way it was consumed via MEF. So I'd suggest merging its content into one of `NuGet.Clients` projects.

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: That sounds like a good idea. Let me search and double check, I thought PM used this, but I hadn't verified that.

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/VisualStudio/ExtensibleSourceRepositoryProvider.cs]: I removed the if def and moved the extensible source provider up to the PM.VS where it is used. Whatever use to use this at a lower level must be gone :smile:

**Comment by emgarten** [comment]
🔔 Moved the extensible source provider to PM.VS, removed MEF assembly references from nuget.protocol since they are no longer needed.

**Review by alpaix — APPROVED** [review]
👍 LGTM

**Comment by mishra14** [comment]
Loc impact: We need to make changes in our loc project and nuget.msbuild file in the TFS repo to reflect the change in dlls/projects. The vendors will then be able to pick this up.
